File: 224-basic-calculator/basic-calculator.java
class Solution {
    public int calculate(String s) {
        int currentNum = 0;
        int result = 0;
        int sign = 1; // old sign value
        Stack<Integer> st = new Stack<>();

        for (Character c: s.toCharArray()) {
            if (c == ' ') {
                continue;
            }
            if (Character.isDigit(c)) {
                currentNum  = currentNum * 10 + (c - 48);
                continue;
            }
            if (c == '+') {
                result += currentNum * sign; // use old sign
                sign = 1;
            }
            else if (c == '-') {
                result += currentNum * sign;
                sign = -1;
            }
            else if (c == '(') {
                st.push(result);
                st.push(sign);
                result = 0;
                // currenNum = 0;
                sign = 1;
            }
            else if (c == ')') {
                result += currentNum * sign;
                result *= st.pop();
                result += st.pop();
            }
            currentNum = 0;
        }
        // System.out.println(currentNum * sign + " " + st);
        result += (currentNum * sign);
        // System.out.println(currentNum * sign + " " + st);
        // int sum = 0;
        // for (int x : st) {
        //     sum += x;
        // }
        return result;
    }
}

File: 224-basic-calculator/basic-calculator.py
class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        sign = 1
        currentNum = 0
        result = 0
        s += '+'
        for i in s:
            if i.isdigit():
                currentNum = currentNum * 10 + int(i)
                continue
            if i == ' ':
                continue
            if i == '+':
                result += currentNum * sign
                sign = 1
            elif i == '-':
                result += currentNum * sign
                sign = -1
            elif i == '(':
                stack.append(result)
                stack.append(sign)
                # print(stack)
                sign = 1
                result = 0
            elif i == ')':
                result += currentNum * sign
                result *= stack.pop()
                result += (stack.pop())
                sign = 1
            currentNum = 0
        return result

            

            
